README.md
[![Go Doc][godoc-image]][godoc-url]
[![Build Status][workflow-image]][workflow-url]
[![Go Report Card][goreport-image]][goreport-url]
[![Test Coverage][coverage-image]][coverage-url]
[![Maintainability][maintainability-image]][maintainability-url]

# flagit

flagit allows you to use the `flag` struct tag on your Go struct fields.
You can then read the values for those fields from the command-line arguments and parse them to the corresponding types of struct fields.

## Quick Start

```go
package main

import (
  "fmt"
  "net/url"
  "time"

  "github.com/moorara/flagit"
)

// Spec is a struct for mapping its fields to command-line flags.
type Spec struct {
  // Flag fields
  Verbose bool `flag:"verbose"`

  // Nested fields
  Options struct {
    Port     uint16 `flag:"port"`
    LogLevel string `flag:"log-level"`
  }

  // Nested fields with prefix
  Config struct {
    Timeout   time.Duration `flag:"timeout"`
    Endpoints []url.URL     `flag:"endpoints"`
  } `flag:"config-"`
}

func main() {
  spec := new(Spec)

  if err := flagit.Populate(spec, false); err != nil {
    panic(err)
  }

  fmt.Printf("%+v\n", spec)
}
```

## Examples

You can find more examples of using `flagit` [here](./examples).

## Documentation

### Supported Types

  - `string`, `*string`, `[]string`
  - `bool`, `*bool`, `[]bool`
  - `float32`, `float64`
  - `*float32`, `*float64`
  - `[]float32`, `[]float64`
  - `int`, `int8`, `int16`, `int32`, `int64`
  - `*int`, `*int8`, `*int16`, `*int32`, `*int64`
  - `[]int`, `[]int8`, `[]int16`, `[]int32`, `[]int64`
  - `uint`, `uint8`, `uint16`, `uint32`, `uint64`
  - `*uint`, `*uint8`, `*uint16`, `*uint32`, `*uint64`
  - `[]uint`, `[]uint8`, `[]uint16`, `[]uint32`, `[]uint64`
  - `url.URL`, `*url.URL`, `[]url.URL`
  - `regexp.Regexp`, `*regexp.Regexp`, `[]regexp.Regexp`
  - `time.Duration`, `*time.Duration`, `[]time.Duration`

The supported syntax for Regexp is [POSIX Regular Expressions](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions).
Nested structs are also supported.


[godoc-url]: https://pkg.go.dev/github.com/moorara/flagit
[godoc-image]: https://godoc.org/github.com/moorara/flagit?status.svg
[workflow-url]: https://github.com/moorara/flagit/actions
[workflow-image]: https://github.com/moorara/flagit/workflows/Main/badge.svg
[goreport-url]: https://goreportcard.com/report/github.com/moorara/flagit
[goreport-image]: https://goreportcard.com/badge/github.com/moorara/flagit
[coverage-url]: https://codeclimate.com/github/moorara/flagit/test_coverage
[coverage-image]: https://api.codeclimate.com/v1/badges/f441152938de958f7ebe/test_coverage
[maintainability-url]: https://codeclimate.com/github/moorara/flagit/maintainability
[maintainability-image]: https://api.codeclimate.com/v1/badges/f441152938de958f7ebe/maintainability


CHANGELOG.md
# Changelog

## [v0.1.3](https://github.com/moorara/flagit/tree/v0.1.3) (2020-10-18)

[Full Changelog](https://github.com/moorara/flagit/compare/v0.1.2...v0.1.3)

**Closed issues:**

- Support Pointer Types [\#15](https://github.com/moorara/flagit/issues/15)

**Merged pull requests:**

- Support pointers [\#16](https://github.com/moorara/flagit/pull/16) ([moorara](https://github.com/moorara))
- Refactor and fix a bug in tests [\#14](https://github.com/moorara/flagit/pull/14) ([moorara](https://github.com/moorara))

## [v0.1.2](https://github.com/moorara/flagit/tree/v0.1.2) (2020-10-16)

[Full Changelog](https://github.com/moorara/flagit/compare/v0.1.1...v0.1.2)

**Merged pull requests:**

- Update README.md [\#13](https://github.com/moorara/flagit/pull/13) ([moorara](https://github.com/moorara))
- Add support for Regexp type [\#12](https://github.com/moorara/flagit/pull/12) ([moorara](https://github.com/moorara))

## [v0.1.1](https://github.com/moorara/flagit/tree/v0.1.1) (2020-10-06)

[Full Changelog](https://github.com/moorara/flagit/compare/v0.1.0...v0.1.1)

**Closed issues:**

- Remove the need for flag tag for nested structs [\#10](https://github.com/moorara/flagit/issues/10)
- Add support for flag description/help text [\#8](https://github.com/moorara/flagit/issues/8)

**Merged pull requests:**

- Remove the need for flag tag for nested structs [\#11](https://github.com/moorara/flagit/pull/11) ([moorara](https://github.com/moorara))
- Add support for help text [\#9](https://github.com/moorara/flagit/pull/9) ([moorara](https://github.com/moorara))

## [v0.1.0](https://github.com/moorara/flagit/tree/v0.1.0) (2020-10-05)

[Full Changelog](https://github.com/moorara/flagit/compare/445043834a88b72efc476715a73e69258a65ddb1...v0.1.0)

**Merged pull requests:**

- Nested structs [\#7](https://github.com/moorara/flagit/pull/7) ([moorara](https://github.com/moorara))
- Implement error handling [\#6](https://github.com/moorara/flagit/pull/6) ([moorara](https://github.com/moorara))
- Finish methods [\#5](https://github.com/moorara/flagit/pull/5) ([moorara](https://github.com/moorara))
- Iteate on the code [\#4](https://github.com/moorara/flagit/pull/4) ([moorara](https://github.com/moorara))
- Add GitHub files [\#3](https://github.com/moorara/flagit/pull/3) ([moorara](https://github.com/moorara))
- Bootstrap 🚀 [\#1](https://github.com/moorara/flagit/pull/1) ([moorara](https://github.com/moorara))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


example_test.go
package flagit_test

import (
	"flag"
	"fmt"
	"net/url"
	"os"
	"time"

	"github.com/moorara/flagit"
)

func ExamplePopulate() {
	// spec is a struct for mapping its fields to command-line flags.
	spec := struct {
		// Flag fields
		Verbose bool `flag:"verbose"`

		// Nested fields
		Options struct {
			Port     uint16 `flag:"port"`
			LogLevel string `flag:"log-level"`
		}

		// Nested fields with prefix
		Config struct {
			Timeout   time.Duration `flag:"timeout"`
			Endpoints []url.URL     `flag:"endpoints"`
		} `flag:"config-"`
	}{}

	if err := flagit.Populate(&spec, false); err != nil {
		panic(err)
	}

	fmt.Printf("%+v\n", spec)
}

func ExampleRegisterFlags() {
	// spec is a struct for mapping its fields to command-line flags.
	spec := struct {
		// Flag fields
		Verbose bool `flag:"verbose,enable verbose logs"`

		// Nested fields
		Options struct {
			Port     uint16 `flag:"port,the port number (1024-65535)"`
			LogLevel string `flag:"log-level,the logging level (debug|info|warn|error)"`
		}

		// Nested fields with prefix
		Config struct {
			Timeout   time.Duration `flag:"timeout,the request timeout"`
			Endpoints []url.URL     `flag:"endpoints,the replica endpoints"`
		} `flag:"config-"`
	}{}

	fs := flag.NewFlagSet("app", flag.ContinueOnError)
	if err := flagit.RegisterFlags(fs, &spec, false); err != nil {
		panic(err)
	}

	if err := fs.Parse(os.Args[1:]); err != nil {
		panic(err)
	}

	fmt.Printf("%+v\n", spec)
}


flagit.go
// Package flagit adds support for a new struct tag: flag.
// You can tag your struct fields with the flag tag and parse command-line arguments into your struct fields.
// Nested structs are also supported. You can either parse the command-line arguments using this package or the built-in flag package.
package flagit

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"reflect"
	"regexp"
	"strings"

	"github.com/moorara/flagit/set"
)

const (
	flagTag = "flag"
	sepTag  = "sep"
)

var (
	flagNameRE = regexp.MustCompile(`^[A-Za-z]([0-9A-Za-z-.]*[0-9A-Za-z])?$`)
	flagArgRE  = regexp.MustCompile("^-{1,2}[A-Za-z]([0-9A-Za-z-.]*[0-9A-Za-z])?")
)

type fieldInfo struct {
	value reflect.Value
	name  string
	flag  string
	help  string
	sep   string
}

// flagValue implements the flag.Value interface.
type flagValue struct {
	continueOnError bool
	value           reflect.Value
	sep             string
}

// String is called for getting and printing the default value.
// Default value is already included in the usage string.
func (v flagValue) String() string {
	return ""
}

func (v flagValue) Set(val string) error {
	if _, err := set.Value(v.value, v.sep, val); err != nil {
		if v.continueOnError {
			return nil
		}
		return err
	}

	return nil
}

func validateStruct(s interface{}) (reflect.Value, error) {
	v := reflect.ValueOf(s) // reflect.Value --> v.Type(), v.Kind(), v.NumField()
	t := reflect.TypeOf(s)  // reflect.Type --> t.Kind(), t.Name(), t.NumField()

	// A pointer to a struct should be passed
	if t.Kind() != reflect.Ptr {
		return reflect.Value{}, errors.New("non-pointer type: you should pass a pointer to a struct type")
	}

	// Navigate to the pointer value
	v = v.Elem()
	t = t.Elem()

	if t.Kind() != reflect.Struct {
		return reflect.Value{}, errors.New("non-struct type: you should pass a pointer to a struct type")
	}

	return v, nil
}

func isStructSupported(t reflect.Type) bool {
	return (t.PkgPath() == "net/url" && t.Name() == "URL") ||
		(t.PkgPath() == "regexp" && t.Name() == "Regexp")
}

func isNestedStruct(t reflect.Type) bool {
	if t.Kind() != reflect.Struct {
		return false
	}

	if isStructSupported(t) {
		return false
	}

	return true
}

func isTypeSupported(t reflect.Type) bool {
	switch t.Kind() {
	case reflect.String:
		return true
	case reflect.Bool:
		return true
	case reflect.Float32, reflect.Float64:
		return true
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return true
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return true
	case reflect.Struct:
		return isStructSupported(t)
	case reflect.Ptr, reflect.Slice:
		return isTypeSupported(t.Elem())
	default:
		return false
	}
}

func getFlagValue(flag string) string {
	flagRegex := regexp.MustCompile("-{1,2}" + flag)

	for i, arg := range os.Args {
		if flagRegex.MatchString(arg) {
			if s := strings.Index(arg, "="); s > 0 {
				return arg[s+1:]
			}

			if i+1 < len(os.Args) {
				if val := os.Args[i+1]; !flagArgRE.MatchString(val) {
					return val
				}
			}

			// For boolean flags
			return "true"
		}
	}

	return ""
}

func iterateOnFields(prefix string, vStruct reflect.Value, continueOnError bool, handle func(f fieldInfo) error) error {
	// Iterate over struct fields
	for i := 0; i < vStruct.NumField(); i++ {
		v := vStruct.Field(i)        // reflect.Value       --> vField.Kind(), vField.Type().Name(), vField.Type().Kind(), vField.Interface()
		t := v.Type()                // reflect.Type        --> t.Kind(), t.PkgPath(), t.Name(), t.NumField()
		f := vStruct.Type().Field(i) // reflect.StructField --> f.Name, f.Type.Name(), f.Type.Kind(), f.Tag.Get(tag)

		// Recursively, iterate on nested structs with flag tag
		if isNestedStruct(t) {
			newPrefix := prefix + f.Tag.Get(flagTag)
			if err := iterateOnFields(newPrefix, v, continueOnError, handle); err != nil {
				return err
			}
		}

		// Skip unexported and unsupported fields
		if !v.CanSet() || !isTypeSupported(t) {
			continue
		}

		// `flag:"..."`
		val := f.Tag.Get(flagTag)
		if val == "" {
			continue
		}

		var flagName, flagHelp string
		if strings.Contains(val, ",") {
			subs := strings.Split(val, ",")
			flagName = subs[0]
			flagHelp = subs[1]
		} else {
			flagName = val
		}

		// Apply prefix
		flagName = prefix + flagName

		// Sanitize the flag name
		if !flagNameRE.MatchString(flagName) {
			if continueOnError {
				continue
			}
			return fmt.Errorf("invalid flag name: %s", flagName)
		}

		// `sep:"..."`
		sep := f.Tag.Get(sepTag)
		if sep == "" {
			sep = ","
		}

		err := handle(fieldInfo{
			value: v,
			name:  f.Name,
			flag:  flagName,
			help:  flagHelp,
			sep:   sep,
		})

		if err != nil {
			return err
		}
	}

	return nil
}

// Populate accepts the pointer to a struct type.
// For those struct fields that have the flag tag, it will read values from command-line flags and parse them to the appropriate types.
// This method does not use the built-in flag package for parsing and reading the flags.
func Populate(s interface{}, continueOnError bool) error {
	v, err := validateStruct(s)
	if err != nil {
		return err
	}

	return iterateOnFields("", v, continueOnError, func(f fieldInfo) error {
		if val := getFlagValue(f.flag); val != "" {
			if _, err := set.Value(f.value, f.sep, val); err != nil {
				if continueOnError {
					return nil
				}
				return err
			}
		}

		return nil
	})
}

// RegisterFlags accepts a flag set and the pointer to a struct type.
// For those struct fields that have the flag tag, it will register a flag on the given flag set.
// The current values of the struct fields will be used as default values for the registered flags.
// Once the Parse method on the flag set is called, the values will be read, parsed to the appropriate types, and assigned to the corresponding struct fields.
func RegisterFlags(fs *flag.FlagSet, s interface{}, continueOnError bool) error {
	v, err := validateStruct(s)
	if err != nil {
		return err
	}

	return iterateOnFields("", v, continueOnError, func(f fieldInfo) error {
		if fs.Lookup(f.flag) != nil {
			if continueOnError {
				return nil
			}
			return fmt.Errorf("flag already registered: %s", f.flag)
		}

		// Create usage string
		var usage string

		if f.help != "" {
			usage = f.help + "\n"
		}

		switch f.value.Kind() {
		case reflect.Slice:
			usage += fmt.Sprintf("%-15s []%s\n%-15s %v\n%-15s %s",
				"data type:", reflect.TypeOf(f.value.Interface()).Elem(),
				"default value:", f.value.Interface(),
				"separator:", f.sep,
			)
		case reflect.Struct:
			usage += fmt.Sprintf("%-15s %s\n%-15s %+v",
				"data type:", f.value.Type(),
				"default value:", f.value.Interface(),
			)
		default:
			usage += fmt.Sprintf("%-15s %s\n%-15s %v",
				"data type:", f.value.Type(),
				"default value:", f.value.Interface(),
			)
		}

		// Register the flag
		switch f.value.Kind() {
		case reflect.Bool:
			// f.value.CanAddr() expected to be true
			// f.value.Addr().Interface().(*bool) expected to be ok
			ptr := f.value.Addr().Interface().(*bool)
			fs.BoolVar(ptr, f.flag, f.value.Bool(), usage)
		default:
			fv := &flagValue{continueOnError, f.value, f.sep}
			fs.Var(fv, f.flag, usage)
		}

		return nil
	})
}


helper.go
package flagit

import (
	"fmt"
	"net/url"
	"reflect"
	"regexp"
	"strconv"
	"time"
)

func setString(v reflect.Value, val string) (bool, error) {
	if v.String() == val {
		return false, nil
	}

	v.SetString(val)
	return true, nil
}

func setBool(v reflect.Value, val string) (bool, error) {
	b, err := strconv.ParseBool(val)
	if err != nil {
		return false, err
	}

	if v.Bool() == b {
		return false, nil
	}

	v.SetBool(b)
	return true, nil
}

func setFloat32(v reflect.Value, val string) (bool, error) {
	f, err := strconv.ParseFloat(val, 32)
	if err != nil {
		return false, err
	}

	if v.Float() == f {
		return false, nil
	}

	v.SetFloat(f)
	return true, nil
}

func setFloat64(v reflect.Value, val string) (bool, error) {
	f, err := strconv.ParseFloat(val, 64)
	if err != nil {
		return false, err
	}

	if v.Float() == f {
		return false, nil
	}

	v.SetFloat(f)
	return true, nil
}

func setInt(v reflect.Value, val string) (bool, error) {
	i, err := strconv.ParseInt(val, 10, 64)
	if err != nil {
		return false, err
	}

	if v.Int() == i {
		return false, nil
	}

	v.SetInt(i)
	return true, nil
}

func setInt8(v reflect.Value, val string) (bool, error) {
	i, err := strconv.ParseInt(val, 10, 8)
	if err != nil {
		return false, err
	}

	if v.Int() == i {
		return false, nil
	}

	v.SetInt(i)
	return true, nil
}

func setInt16(v reflect.Value, val string) (bool, error) {
	i, err := strconv.ParseInt(val, 10, 16)
	if err != nil {
		return false, err
	}

	if v.Int() == i {
		return false, nil
	}

	v.SetInt(i)
	return true, nil
}

func setInt32(v reflect.Value, val string) (bool, error) {
	i, err := strconv.ParseInt(val, 10, 32)
	if err != nil {
		return false, err
	}

	if v.Int() == i {
		return false, nil
	}

	v.SetInt(i)
	return true, nil
}

func setInt64(v reflect.Value, val string) (bool, error) {
	if t := v.Type(); t.PkgPath() == "time" && t.Name() == "Duration" {
		d, err := time.ParseDuration(val)
		if err != nil {
			return false, err
		}

		if v.Interface() == d {
			return false, nil
		}

		v.Set(reflect.ValueOf(d))
		return true, nil
	}

	i, err := strconv.ParseInt(val, 10, 64)
	if err != nil {
		return false, err
	}

	if v.Int() == i {
		return false, nil
	}

	v.SetInt(i)
	return true, nil
}

func setUint(v reflect.Value, val string) (bool, error) {
	// uint size and range are platform-dependent
	u, err := strconv.ParseUint(val, 10, 64)
	if err != nil {
		return false, err
	}

	if v.Uint() == u {
		return false, nil
	}

	v.SetUint(u)
	return true, nil
}

func setUint8(v reflect.Value, val string) (bool, error) {
	u, err := strconv.ParseUint(val, 10, 8)
	if err != nil {
		return false, err
	}

	if v.Uint() == u {
		return false, nil
	}

	v.SetUint(u)
	return true, nil
}

func setUint16(v reflect.Value, val string) (bool, error) {
	u, err := strconv.ParseUint(val, 10, 16)
	if err != nil {
		return false, err
	}

	if v.Uint() == u {
		return false, nil
	}

	v.SetUint(u)
	return true, nil
}

func setUint32(v reflect.Value, val string) (bool, error) {
	u, err := strconv.ParseUint(val, 10, 32)
	if err != nil {
		return false, err
	}

	if v.Uint() == u {
		return false, nil
	}

	v.SetUint(u)
	return true, nil
}

func setUint64(v reflect.Value, val string) (bool, error) {
	u, err := strconv.ParseUint(val, 10, 64)
	if err != nil {
		return false, err
	}

	if v.Uint() == u {
		return false, nil
	}

	v.SetUint(u)
	return true, nil
}

func setStruct(v reflect.Value, val string) (bool, error) {
	t := v.Type()

	if t.PkgPath() == "net/url" && t.Name() == "URL" {
		u, err := url.Parse(val)
		if err != nil {
			return false, err
		}

		// u is a pointer
		if reflect.DeepEqual(v.Interface(), *u) {
			return false, nil
		}

		// u is a pointer
		v.Set(reflect.ValueOf(u).Elem())
		return true, nil
	} else if t.PkgPath() == "regexp" && t.Name() == "Regexp" {
		r, err := regexp.CompilePOSIX(val)
		if err != nil {
			return false, err
		}

		// r is a pointer
		if reflect.DeepEqual(v.Interface(), *r) {
			return false, nil
		}

		// r is a pointer
		v.Set(reflect.ValueOf(r).Elem())
		return true, nil
	}

	return false, fmt.Errorf("unsupported type: %s.%s", t.PkgPath(), t.Name())
}

func setStringSlice(v reflect.Value, vals []string) (bool, error) {
	if reflect.DeepEqual(v.Interface(), vals) {
		return false, nil
	}

	v.Set(reflect.ValueOf(vals))
	return true, nil
}

func setBoolSlice(v reflect.Value, vals []string) (bool, error) {
	bools := []bool{}
	for _, val := range vals {
		b, err := strconv.ParseBool(val)
		if err != nil {
			return false, err
		}

		bools = append(bools, b)
	}

	if reflect.DeepEqual(v.Interface(), bools) {
		return false, nil
	}

	v.Set(reflect.ValueOf(bools))
	return true, nil
}

func setFloat32Slice(v reflect.Value, vals []string) (bool, error) {
	floats := []float32{}
	for _, val := range vals {
		f, err := strconv.ParseFloat(val, 32)
		if err != nil {
			return false, err
		}

		floats = append(floats, float32(f))
	}

	if reflect.DeepEqual(v.Interface(), floats) {
		return false, nil
	}

	v.Set(reflect.ValueOf(floats))
	return true, nil
}

func setFloat64Slice(v reflect.Value, vals []string) (bool, error) {
	floats := []float64{}
	for _, val := range vals {
		f, err := strconv.ParseFloat(val, 64)
		if err != nil {
			return false, err
		}

		floats = append(floats, f)
	}

	if reflect.DeepEqual(v.Interface(), floats) {
		return false, nil
	}

	v.Set(reflect.ValueOf(floats))
	return true, nil
}

func setIntSlice(v reflect.Value, vals []string) (bool, error) {
	// int size and range are platform-dependent
	ints := []int{}
	for _, val := range vals {
		i, err := strconv.ParseInt(val, 10, 64)
		if err != nil {
			return false, err
		}

		ints = append(ints, int(i))
	}

	if reflect.DeepEqual(v.Interface(), ints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(ints))
	return true, nil
}

func setInt8Slice(v reflect.Value, vals []string) (bool, error) {
	ints := []int8{}
	for _, val := range vals {
		i, err := strconv.ParseInt(val, 10, 8)
		if err != nil {
			return false, err
		}

		ints = append(ints, int8(i))
	}

	if reflect.DeepEqual(v.Interface(), ints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(ints))
	return true, nil
}

func setInt16Slice(v reflect.Value, vals []string) (bool, error) {
	ints := []int16{}
	for _, val := range vals {
		i, err := strconv.ParseInt(val, 10, 16)
		if err != nil {
			return false, err
		}

		ints = append(ints, int16(i))
	}

	if reflect.DeepEqual(v.Interface(), ints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(ints))
	return true, nil
}

func setInt32Slice(v reflect.Value, vals []string) (bool, error) {
	ints := []int32{}
	for _, val := range vals {
		i, err := strconv.ParseInt(val, 10, 32)
		if err != nil {
			return false, err
		}

		ints = append(ints, int32(i))
	}

	if reflect.DeepEqual(v.Interface(), ints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(ints))
	return true, nil
}

func setInt64Slice(v reflect.Value, vals []string) (bool, error) {
	t := reflect.TypeOf(v.Interface()).Elem()

	if t.PkgPath() == "time" && t.Name() == "Duration" {
		durations := []time.Duration{}
		for _, val := range vals {
			d, err := time.ParseDuration(val)
			if err != nil {
				return false, err
			}

			durations = append(durations, d)
		}

		if reflect.DeepEqual(v.Interface(), durations) {
			return false, nil
		}

		v.Set(reflect.ValueOf(durations))
		return true, nil
	}

	ints := []int64{}
	for _, val := range vals {
		i, err := strconv.ParseInt(val, 10, 64)
		if err != nil {
			return false, err
		}

		ints = append(ints, i)
	}

	if reflect.DeepEqual(v.Interface(), ints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(ints))
	return true, nil
}

func setUintSlice(v reflect.Value, vals []string) (bool, error) {
	// uint size and range are platform-dependent
	uints := []uint{}
	for _, val := range vals {
		u, err := strconv.ParseUint(val, 10, 64)
		if err != nil {
			return false, err
		}

		uints = append(uints, uint(u))
	}

	if reflect.DeepEqual(v.Interface(), uints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(uints))
	return true, nil
}

func setUint8Slice(v reflect.Value, vals []string) (bool, error) {
	uints := []uint8{}
	for _, val := range vals {
		u, err := strconv.ParseUint(val, 10, 8)
		if err != nil {
			return false, err
		}

		uints = append(uints, uint8(u))
	}

	if reflect.DeepEqual(v.Interface(), uints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(uints))
	return true, nil
}

func setUint16Slice(v reflect.Value, vals []string) (bool, error) {
	uints := []uint16{}
	for _, val := range vals {
		u, err := strconv.ParseUint(val, 10, 16)
		if err != nil {
			return false, err
		}

		uints = append(uints, uint16(u))
	}

	if reflect.DeepEqual(v.Interface(), uints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(uints))
	return true, nil
}

func setUint32Slice(v reflect.Value, vals []string) (bool, error) {
	uints := []uint32{}
	for _, val := range vals {
		u, err := strconv.ParseUint(val, 10, 32)
		if err != nil {
			return false, err
		}

		uints = append(uints, uint32(u))
	}

	if reflect.DeepEqual(v.Interface(), uints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(uints))
	return true, nil
}

func setUint64Slice(v reflect.Value, vals []string) (bool, error) {
	uints := []uint64{}
	for _, val := range vals {
		u, err := strconv.ParseUint(val, 10, 64)
		if err != nil {
			return false, err
		}

		uints = append(uints, u)
	}

	if reflect.DeepEqual(v.Interface(), uints) {
		return false, nil
	}

	v.Set(reflect.ValueOf(uints))
	return true, nil
}

func setStructSlice(v reflect.Value, vals []string) (bool, error) {
	t := reflect.TypeOf(v.Interface()).Elem()

	if t.PkgPath() == "net/url" && t.Name() == "URL" {
		urls := []url.URL{}
		for _, val := range vals {
			u, err := url.Parse(val)
			if err != nil {
				return false, err
			}

			urls = append(urls, *u)
		}

		// []url.URL
		if reflect.DeepEqual(v.Interface(), urls) {
			return false, nil
		}

		v.Set(reflect.ValueOf(urls))
		return true, nil
	} else if t.PkgPath() == "regexp" && t.Name() == "Regexp" {
		regexps := []regexp.Regexp{}
		for _, val := range vals {
			r, err := regexp.CompilePOSIX(val)
			if err != nil {
				return false, err
			}

			regexps = append(regexps, *r)
		}

		// []regexp.Regexp
		if reflect.DeepEqual(v.Interface(), regexps) {
			return false, nil
		}

		v.Set(reflect.ValueOf(regexps))
		return true, nil
	}

	return false, fmt.Errorf("unsupported type: %s.%s", t.PkgPath(), t.Name())
}


examples/2-RegisterFlags/main.go
package main

import (
	"flag"
	"fmt"
	"net/url"
	"os"
	"time"

	"github.com/moorara/flagit"
)

// Spec is a struct for mapping its fields to command-line flags.
type Spec struct {
	// Flag fields
	Verbose bool `flag:"verbose,enable verbose logs"`

	// Nested fields
	Options struct {
		Port     uint16 `flag:"port,the port number (1024-65535)"`
		LogLevel string `flag:"log-level,the logging level (debug|info|warn|error)"`
	}

	// Nested fields with prefix
	Config struct {
		Timeout   time.Duration `flag:"timeout,the request timeout"`
		Endpoints []url.URL     `flag:"endpoints,the replica endpoints"`
	} `flag:"config-"`
}

func main() {
	spec := new(Spec)
	fs := flag.NewFlagSet("app", flag.ContinueOnError)

	if err := flagit.RegisterFlags(fs, spec, false); err != nil {
		panic(err)
	}

	if err := fs.Parse(os.Args[1:]); err != nil {
		panic(err)
	}

	fmt.Printf("%+v\n", spec)
}
